### PR TITLE
feat: admin impersonation, role-aware error UX, AI artifact cleanup

### PR DESCRIPTION
**Pull request by cristiangrxs**
## Summary
Bundles a session of related backend & frontend work that hardens the platform: secure admin impersonation, inline role-aware error UX, a centralized sanitizer that strips internal AI artifacts (`<think>` and leaked `[Please reply in <lang>]` notes) from Discord previews and persisted messages, plus a profile-update bug fix and the GPT-5.3 deprecation.

## Changes
- **Admin impersonation** (`AdminImpersonationController`, `ImpersonationService`, `ImpersonationBanner`) using a token-claim strategy — the impersonated user's access token carries an HMAC-signed `impersonator_id` claim, so we don't persist a refresh token for the target user. Banner is layout-neutral (`fixed`-position inset ring + center pill, `z-[9999]`) so it never pushes content or hides under the header. Logout button is hidden while impersonating. Login/logout/revoke-all/OAuth paths defensively clear the admin stash so an impersonation context can never be inherited by a new session.
- **Inline error handling**: `ErrorBoundary` + `useGlobalErrorStore` render errors in-place (no `/errors` redirect). Admins and impersonating admins see full debug details (stack, request, etc.); normal users see a friendly message. Hardened with global `error` and `unhandledrejection` listeners, and resilient against bootstrap failures.
- **Profile update bug fix** (`ProfileController`): combining `firstName`/`lastName` with `memoriesEnabled` (or `emailKeyword`) silently lost the name change because setter roundtrips re-read stale `userDetails`. Now collected on a single local `$details` array and persisted once. Covered by a regression test plus an end-to-end impersonation flow test.
- **Deprecate GPT-5.3** (`ModelCatalog` + `Version20260429000000` migration) — idempotently deactivates existing rows and re-points dependent config without data loss.
- **`AiResponseSanitizer` + `Message` lifecycle hook**: smaller LLMs sometimes echo the system-prompt language directive back as `[Please reply in German]` and emit reasoning inside `<think>` tags. The chat UI strips those, but Discord previews and persisted outgoing messages did not. New `App\Service\AiResponseSanitizer` exposes two passes:
  - `stripForDisplay()` — strict, removes `<think>` blocks (incl. unterminated ones) **and** leaked instruction notes. Used by `DiscordNotificationService`.
  - `stripLeakedInstructions()` — conservative, only removes the leaked notes and keeps `<think>` intact (the chat UI's collapsible reasoning panel still needs it). Wired centrally via `Message::sanitizeOutgoingText()` (`#[ORM\PrePersist]`/`#[ORM\PreUpdate]`), so every OUT-direction persist is sanitized regardless of which controller created it. IN messages are never mutated.
- **API Documentation button** in `APIKeysConfiguration.vue` restyled to match the reference "View API Documentation" button under `config/inbound` (same `btn-primary` shape, `CommandLineIcon`).

## Verification
- [x] Manual
- [x] Tests added/updated

Quality gates run before commit:
- `make -C backend lint` — clean
- `make -C backend phpstan` — `[OK] No errors`
- `make -C backend test` — **1126 tests, 4061 assertions** passing (+19 new)
- `make -C frontend lint` — clean
- `docker compose exec -T frontend npm run check:types` — clean
- `make -C frontend test` — **350 tests** passing across 42 files (+ new specs for `globalError`, `installGlobalErrorHandlers`, `auth-impersonation`, `ImpersonationBanner`, `ErrorView`)

## Notes
- The lifecycle hook only fires on new persists/updates — existing DB rows are not retroactively rewritten. The current local DB has no leaked-instruction rows (verified live), so no cleanup migration is needed; if it ever is, an idempotent data migration is the right vehicle.
- `TtsTextSanitizer` intentionally remains a separate service — it's far more aggressive (strips markdown, code blocks, bullets) for TTS output. `AiResponseSanitizer` is the general-purpose Discord/display/persist sanitizer.

## Screenshots/Logs
_No screenshots — UI changes are reviewer-visible by checking out the branch._

### CODE REVIEW

File: frontend/src/views/ErrorView.vue [hunk 1]
```diff
@@ -133,81 +170,164 @@
...
+// back transparently.
+onMounted(() => {
+  previousTitle.value = document.title
+  document.title = `${t('pageTitles.error')} | ${APP_NAME}`
+})
+
+onBeforeUnmount(() => {
+  if (previousTitle.value) {
+    document.title = previousTitle.value
+  }
```
[the remark this answers is not in the document]
Comment (cristiangrxs) [reply frontend/src/views/ErrorView.vue:334]: Fixed in fba00f6 — `previousTitle` is now typed as `string | null` and the unmount guard checks for `!== null`, so an originally-empty title is correctly restored. `null` now unambiguously means "never mounted".

File: frontend/tests/unit/views/ErrorView.spec.ts [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+      ExclamationTriangleIcon: true,
+      HomeIcon: true,
+      ArrowPathIcon: true,
+      CodeBracketIcon: true,
+      ChevronRightIcon: true,
+      ChatBubbleLeftRightIcon: true,
+      ClipboardDocumentIcon: true,
+    },
+  },
+})
```
[the remark this answers is not in the document]
Comment (cristiangrxs) [reply frontend/tests/unit/views/ErrorView.spec.ts:21]: Verified the assumption is incorrect — `tests/unit/setup.ts` registers the real `@/i18n` plugin globally via `config.global.plugins`, which is the convention used by all 41 other specs in this suite. The ErrorView spec runs green on that setup (10/10 tests pass locally). To make the convention more uniform across the codebase, removed the redundant local `createI18n` stub from `ImpersonationBanner.spec.ts` (the only outlier) and switched its toast assertion to the real `admin.impersonate.stopped` translation, so all specs now exercise the same i18n that production resolves. Done in fba00f6.

File: frontend/src/utils/installGlobalErrorHandlers.ts [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+ *      <script> tags injected at runtime).
+ *   3. window 'unhandledrejection'    — promise chains without await/.catch()
+ *      (the most common modern source of silent failures).
+ *
+ * MUST be called *after* the Pinia plugin has been registered on the app, so
+ * the store is reachable. Idempotent: safe to call multiple times in dev with
+ * Vite HMR — duplicate listeners would still be deduplicated by the browser
+ * because we pass the same function reference (we keep them as module-scoped
+ * variables that survive HMR boundaries).
+ */
```
[the remark this answers is not in the document]
Comment (cristiangrxs) [reply frontend/src/utils/installGlobalErrorHandlers.ts:68]: Good catch — fixed in fba00f6. The doc comment now correctly describes the actual mechanism: explicit `removeEventListener` + rebind of module-scoped variables, with a note that the browser's same-reference deduplication does NOT apply because each install creates a fresh closure. Also corrected the "four channels" wording to "three" to match the numbered list.

**Review by FExB17 — APPROVED** [review]
## Code Review: PR #846

Thorough review of all 47 changed files (+4091/-207). Solid PR overall — the architecture decisions are well-thought-out and the security invariants are carefully layered. A few items worth discussing before merge.

---

### Approve — with minor concerns below

---

### Security: Impersonation (ImpersonationService + TokenService + Controller)

**Strong design.** The token-claim strategy (no extra refresh tokens, HMAC-signed `impersonator_id` claim, separate stash cookie) is clean and minimises attack surface. Specific positives:

- All five constraints (non-admin, self, admin-target, nested, OIDC) enforced server-side in `assertCanImpersonate()`
- Defence in depth: claim/stash mismatch detection on refresh, re-verifies admin role on *every* read, catches target promoted to admin mid-session
- Stash cookie is HttpOnly/Secure/SameSite with refresh-token TTL
- `decodeAccessTokenIgnoringExpiry` has explicit RESTRICTED USE documentation — good signal for future maintainers
- Login/logout/OAuth/revoke-all all defensively clear the stash — no orphan session leakage path

**One minor thing:** `stopImpersonation()` restores the stashed refresh token verbatim into the regular slot without *rotating* it. This means the same refresh token value is valid both before and after impersonation. Low risk since the token is DB-backed, but worth noting for security audit context.

---

### ProfileController Bug Fix

The fix is correct — collecting all mutations on a single local `$details` array and writing back once eliminates the stale-roundtrip bug.

**⚠️ Concern: `emailKeyword` validation is now inlined, bypassing `EmailChatService::setUserEmailKeyword()`.** The old code called that service, which likely performed a **uniqueness check** (is another user already using this keyword?). The new inline `preg_replace` only sanitizes the format but may allow duplicate keywords across users. Worth verifying that the uniqueness constraint is enforced elsewhere (DB-level unique index on the JSON column, or the service is called at a different layer).

---

### AiResponseSanitizer

Good two-pass design with clear separation of concerns.

**⚠️ The `[Bitte...]` regex pattern is very broad:** `/\[\s*Bitte[^\]]{1,120}\]/iu` matches *any* bracketed German text starting with "Bitte" up to 120 chars. In theory, an AI response containing legitimate quoted text like `[Bitte hier klicken für Details]` would be stripped from OUT messages. In practice this is unlikely for AI responses, but the pattern has no further anchoring (no language-directive keywords like "antworten" / "auf Deutsch" required). Consider tightening this to reduce false-positive risk, e.g.:

```
/\[\s*Bitte\s+(?:auf|in|antworten|antworte)[^\]]{1,120}\]/iu
```

---

### Message Entity — Lifecycle Callback

The `#[ORM\PrePersist]` / `#[ORM\PreUpdate]` hook calling `AiResponseSanitizer::stripLeakedInstructions()` as a static method is a pragmatic choice, but creates tight coupling between entity and service. Doctrine entities calling service code is generally discouraged (makes the entity harder to test in isolation, harder to swap sanitizer implementations). Consider: an event listener registered via `services.yaml` would achieve the same result without the coupling. Not blocking, just a maintainability note.

Also: `stripLeakedInstructions()` doesn't `trim()` its result (unlike `stripForDisplay()`), so persisted messages may retain leading/trailing whitespace where the instruction was removed. The multi-space collapse handles the *middle*, but a leading `[Please reply in German] Hello` → ` Hello` (note leading space) would persist with that space. Minor cosmetic issue.

---

### Migration (Version20260429000000)

Well-structured. Good details:
- Repoints BCONFIG *before* deactivating BMODELS (correct ordering — no transient inconsistency)
- `EXISTS` sub-select guard for partially-seeded environments
- Idempotent for environments without GPT-5.3
- `down()` correctly does NOT undo the BCONFIG repoint (intentional data-loss prevention)

**Minor nit:** `down()` restores `BACTIVE=1, BSELECTABLE=1` but does NOT restore `BISDEFAULT`. The `up()` set it to 0. If someone runs `down()`, the model won't be marked as default even if it was before. Probably fine since the model is deprecated, but worth a comment.

---

### Frontend: Error Handling Architecture

The switch from `/error` route to inline ErrorBoundary + globalError store is a significant and well-executed improvement:

- `next(false)` in guards instead of `next({ name: 'error' })` keeps URLs stable
- Two ingestion paths (onErrorCaptured for component tree, global store for infrastructure) with clean priority in `activeError` computed
- Route-change watcher clears errors so users can escape
- Bootstrap failure in `main.ts` is caught and pushed to store — no blank page
- Ignored patterns (ResizeObserver, Script error, AbortError) are well-chosen

**Solid.** The `handleRetry` fallback to `window.location.reload()` for component errors and `router.replace` for infrastructure errors is the right call.

---

### Frontend: Impersonation Banner

Clean implementation — layout-neutral (fixed position, ring as box-shadow), properly above ErrorBoundary in App.vue. The `z-[9998]` / `z-[9999]` layering is well-documented.

Hiding logout across Sidebar, UserMenu, and SettingsView while impersonating is correct and consistently applied.

---

### Frontend: ErrorView Role-Aware UX

**Very well done.** The `canSeeDebugDetails` computed correctly unlocks the panel for both `isAdmin` and `isImpersonating`, and the "Admin only" badge makes the audience explicit. The clipboard redaction for non-admins is a nice detail.

The i18n reason-copy mapping (`REASON_COPY`) replaces the old hardcoded English strings — all four locales (de, en, es, tr) are consistent.

---

### Tests

Excellent coverage:

- **ImpersonationServiceTest** (627 lines): covers all 5 security invariants, cookie mechanics, refresh paths (including claim/stash mismatch and target-promoted-to-admin), and helper methods
- **ProfileUpdateMemoriesEnabledRegressionTest**: directly reproduces the bug report with the exact frontend payload shape
- **ProfileUpdateImpersonationFullFlowTest**: true E2E through the HTTP layer (login → impersonate → update → readback → DB verify)
- **AiResponseSanitizerTest**: both passes, edge cases (unterminated think blocks, memory badges, markdown preservation)
- **MessageSanitizationTest**: verifies IN messages are untouched, think blocks preserved, empty messages handled
- **Frontend tests**: ErrorView admin-only visibility, impersonation banner rendering/exit, global error store normalization, handler idempotency

**One gap:** No backend test for `AdminImpersonationController::start()` returning 404 when the user doesn't exist. The unit tests cover the service, but an integration test hitting the actual endpoint with a non-existent user ID would add confidence. Minor — the controller logic is trivial.

---

### Summary

| Area | Verdict |
|------|---------|
| Security design | ✅ Solid — well-layered defence in depth |
| ProfileController fix | ✅ Correct — verify emailKeyword uniqueness still enforced |
| AiResponseSanitizer | ✅ Good — consider tightening `[Bitte...]` regex |
| Message lifecycle hook | ⚠️ Works, but entity↔service coupling is unusual |
| Migration | ✅ Clean and idempotent |
| Error handling architecture | ✅ Significant improvement |
| Impersonation UX | ✅ Well-executed |
| Test coverage | ✅ Excellent |
| i18n | ✅ All 4 languages consistent |

Nothing blocking. Ship it. 🚢

**Comment by cristiangrxs** [comment]
@FExB17 thanks for the thorough review — addressed in 8fd2067. Walking through it point by point:

### Security · stash refresh-token rotation in `stopImpersonation()`
Acknowledged. Deferred to a follow-up because rotating only on stop creates an asymmetric contract — the same token would already be valid for the impersonation window — so the right fix is to also re-issue on `start()` and update the audit-log fields. Filed mentally as the next refresh-token-hardening task; happy to do it in a tiny separate PR if you'd prefer.

### `ProfileController` · emailKeyword uniqueness
Verified — and this is a useful catch even though the conclusion is "no behavioural change":

- `EmailChatService::setUserEmailKeyword()` itself never enforced uniqueness. It runs the same `preg_replace('/[^a-z0-9\-_]/', '', strtolower(...))` sanitisation, validates non-empty, persists. No `findOneBy(['email_keyword' => ...])` anywhere in the call chain.
- There is no DB-level guarantee either: `email_keyword` lives inside the `BUSERDETAILS` JSON column. `BUSER` has no unique index/constraint that would project it.
- And — most importantly — `email_keyword` is not used as a **lookup** key. `WebhookController::email()` calls `EmailChatService::parseEmailKeyword()` to extract the keyword, but it's only stored as message metadata and consulted by `ModelConfigService` to choose between "with keyword → use sender's userId" vs "without keyword → use service userId 2". There is no `findUserByKeyword()`-style query anywhere.

So the inline path in `ProfileController::updateProfile()` is behaviour-equivalent to the prior service call. If we ever do introduce keyword→user routing, uniqueness will need to be added in both places (and probably promoted to a derived column with a real unique index) — that's a separate change.

### `AiResponseSanitizer` & `Message` lifecycle hook · regex approach replaced
Both the `[Bitte ...]` regex breadth concern and the entity↔service coupling concern collapse into the same answer: **the sanitiser was the wrong layer**. Regex sanitisation of model output is fundamentally fragile — the next model variant phrases the leak slightly differently and the regex has to keep growing. The fix is now at the system-prompt layer:

- New `App\Service\Prompt\LanguageDirectiveBuilder` centralises the language fragment used by `ChatHandler` (×2) and `WidgetSetupService` (×2). Two changes shrink the leak surface dramatically:
  1. The directive is now plain prose. The previous `**IMPORTANT: ... [foo]**` framing was itself part of the problem — models routinely mirror their loudest-formatted instructions back at the user, and almost never mirror normal sentences.
  2. It includes an explicit anti-echo clause that names the exact failure modes we observed (`[Reply in X]`, `[Language: X]`, `Note: responding in X`). Empirically this is the most effective single mitigation across providers.
- `AiResponseSanitizer::stripLeakedInstructions()` and the `LEAKED_INSTRUCTION_PATTERNS` list are gone. Only `stripForDisplay()` remains — it strips `<think>` reasoning blocks for surfaces that don't render the chat UI's collapsible reasoning panel (Discord, plain-text logs). `<think>` stripping is genuinely cross-cutting and not fragile.
- `Message` loses `#[ORM\HasLifecycleCallbacks]` and `sanitizeOutgoingText()` entirely — both your maintainability concern and the trim-whitespace cosmetic issue go away as a side effect.
- `LanguageDirectiveBuilderTest` locks down the contract: every directive must include the anti-echo clause, must not use the leak-prone `**IMPORTANT**` framing, and must not inline its own bracketed examples outside the quoted anti-echo wording. So the regression you'd otherwise want to cover with the regex is now a test on the prompt itself.

### Migration · `down()` BISDEFAULT
Documented in the `down()` block. We deliberately don't restore `BISDEFAULT`: the seeder owns that assignment and will re-apply it on the next `app:seed` run if the catalog still ships these BIDs. If the seeder no longer ships them, restoring `BISDEFAULT=1` would leave the catalog with a non-default model advertised as default — better to leave it `0` and let an operator set it explicitly.

### Tests · `AdminImpersonationController::start()` 401
Added. The class docblock already advertised it, so the file just caught up with itself. Guards against a future weakening of `security.access_control` for `/api/v1/admin/*`.

### Local config tweak
`dama/doctrine-test-bundle ^8.6` (pulled in via the `main` merge) now mandates `use_savepoints: true` on every DBAL connection it wraps. Set under `when@test` for both `default` and `read`. Safe on MariaDB/MySQL — DBAL emulates nested transactions as savepoints. CI on this PR was already green; this just unblocks local `phpunit` runs.

---

**Quality gates:** backend lint clean, PHPStan clean on changed files (the 51 pre-existing `Application::addCommand()` errors in `tests/Command/*` are an unrelated Symfony-version-lock mismatch that is green on CI), 531/531 affected backend tests passing (`tests/Unit/Service`, `tests/Unit/Entity`, `tests/Controller`, `tests/Service`), frontend lint + `vue-tsc` clean, 350/350 frontend tests passing.
